compiler/transforms/convert_acc_to_csr.py

```diff
@@ -0,0 +1,273 @@
+                llvm.InlineAsmOp(
+                    "csrw $0, $1",
+                    # I = any 12 bit immediate, K = any 5 bit immediate
+                    # The K allows LLVM to emit an `csrrwi` instruction,
+                    # which has room for one 5 bit immediate only.
+                    "I, K",
+                    [addr_val, val],
+                    has_side_effects=True,
+                ),
+            ],
+            [op.state],
+            safe_erase=False,
+        )
+
+
+class LowerAccAwaitToCsr(LowerAccBasePattern):
+    """
+    Lower await ops to a set of assembly that lowers to a buffer.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: acc.AwaitOp, rewriter: PatternRewriter, /):
+        assert isinstance(op.token.type, acc.StateType | acc.TokenType)
+        acc_op = self.get_acc(op.token.type.accelerator)
+
+        # TODO: this is a temporary solution that will be reworked eventually
+
+        # emit a snax_hwpe-style barrier
+        rewriter.replace_matched_op(
+            [
+                # Infinite while loop that unbreaks whenever
+                While(
+                    [],
+                    [],
+                    [
+                        barrier := arith.Constant(acc_op.barrier),
+                        zero := arith.Constant(
+                            builtin.IntegerAttr.from_int_and_width(0, 32)
+                        ),
+                        status := llvm.InlineAsmOp(
+                            "csrr $0, $1",
+                            # I = any 12 bit immediate
+                            # =r = store result in A 32- or 64-bit
+                            # general-purpose register (depending on the platform XLEN)
+                            "=r, I",
+                            [barrier],
+                            [i32],
+                            has_side_effects=True,
+                        ),
+                        # check if not equal to zero
+                        comparison := arith.Cmpi(status, zero, "ne"),
+                        Condition(comparison.results[0]),
+                    ],
+                    [
+                        Yield(),
+                    ],
+                ),
+                addr_val := arith.Constant(builtin.IntegerAttr(965, 12)),  # 0x3c5 = 965
+                zero := arith.Constant(builtin.IntegerAttr.from_int_and_width(0, 5)),
+                llvm.InlineAsmOp(
+                    "csrw $0, $1",
+                    "I, r",
+                    [addr_val, zero],
+                    has_side_effects=True,
+                ),
+                # Three nops for random but important reasons
+                llvm.InlineAsmOp("nop", "", [], [], has_side_effects=True),
+                llvm.InlineAsmOp("nop", "", [], [], has_side_effects=True),
+                llvm.InlineAsmOp("nop", "", [], [], has_side_effects=True),
```

Comment:
what is the reason here?

Reply:
I don't know hahaha @rgantonio ?
Code in question is here: https://github.com/KULeuven-MICAS/snitch_cluster/blob/8d7dff1f33fd9705bfcf1bbcebba246a5ccac25d/target/snitch_cluster/sw/snax/mac/src/mac.c#L16